File: kanbanDLL.java
class Action {
    String task;
    String from;
    String to;

    Action(String task, String from, String to) {
        this.task = task;
        this.from = from;
        this.to = to;
    }

    void apply() {
        System.out.println("APPLY: " + task + " " + from + " → " + to);
    }

    void undo() {
        System.out.println("UNDO: " + task + " " + to + " → " + from);
    }
}

class Node {
    Action action;
    Node prev, next;

    Node(Action action) {
        this.action = action;
    }
}

class History {
    Node head = null;
    Node current = null;

    void addAction(Action action) {
        Node node = new Node(action);

        // Clear future history
        if (current != null && current.next != null) {
            current.next = null;
        }

        if (head == null) {
            head = node;
        } else {
            node.prev = current;
            if (current != null)
                current.next = node;
        }

        current = node;
        action.apply();
    }

    void undo() {
        if (current == null) return;
        current.action.undo();
        current = current.prev;
    }

    void redo() {
        if (current != null && current.next != null) {
            current = current.next;
            current.action.apply();
        }
    }
}

public class kanbanDLL{
    public static void main(String[] args) {
        History history = new History();

        history.addAction(new Action("Task1", "TODO", "IN_PROGRESS"));
        history.addAction(new Action("Task1", "IN_PROGRESS", "DONE"));

        history.undo();
        history.redo();
    }
}

File: kanban_dll.cpp
#include <iostream>
using namespace std;

struct Action {
    string task, from, to;

    void apply() {
        cout << "APPLY: " << task << " " << from << " → " << to << endl;
    }

    void undo() {
        cout << "UNDO: " << task << " " << to << " → " << from << endl;
    }
};

struct Node {
    Action action;
    Node* prev;
    Node* next;

    Node(Action a) : action(a), prev(nullptr), next(nullptr) {}
};

class History {
    Node* head = nullptr;
    Node* current = nullptr;

public:
    void add(Action a) {
        Node* node = new Node(a);

        if (current && current->next) {
            current->next = nullptr;
        }

        if (!head) {
            head = node;
        } else {
            node->prev = current;
            if (current)
                current->next = node;
        }

        current = node;
        a.apply();
    }

    void undo() {
        if (!current) return;
        current->action.undo();
        current = current->prev;
    }

    void redo() {
        if (current && current->next) {
            current = current->next;
            current->action.apply();
        }
    }
};

int main() {
    History history;

    history.add({"Task1", "TODO", "IN_PROGRESS"});
    history.add({"Task1", "IN_PROGRESS", "DONE"});

    history.undo();
    history.redo();

    return 0;
}
